features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5XPalGattClient.h

```diff
@@ -0,0 +1,257 @@
+    );
+
+    /**
+     * see pal::GattClient::write_attribute .
+     */
+    virtual ble_error_t write_attribute(
+        connection_handle_t connection_handle,
+        attribute_handle_t attribute_handle,
+        const ArrayView<const uint8_t>& value
+    );
+
+    /**
+     * see pal::GattClient::queue_prepare_write .
+     */
+    virtual ble_error_t queue_prepare_write(
+        connection_handle_t connection_handle,
+        attribute_handle_t characteristic_value_handle,
+        const ArrayView<const uint8_t>& value,
+        uint16_t offset
+    );
+
+    /**
+     * see pal::GattClient::execute_write_queue .
+     */
+    virtual ble_error_t execute_write_queue(
+        connection_handle_t connection_handle,
+        bool execute
+    );
+
+    // singleton of the ARM Cordio client
+    static nRF5XGattClient& get_client();
+
+    /**
+     * Function call from btle.cpp
+     *
+     * Do not call directly.
+     */
+    static void handle_events(const ble_evt_t *p_ble_evt);
+
+    /**
+     * Called by btle.cpp when a disconnection happens.
+     */
+    static void handle_connection_termination(connection_handle_t connection);
+
+private:
+    struct GattProcedure;
+    struct RegularGattProcedure;
+    struct DiscoverPrimaryServiceProcedure;
+    struct DiscoverPrimaryServiceByUUIDProcedure;
+    struct FindIncludedServicesProcedure;
+    struct DiscoverCharacteristicsProcedure;
+    struct DiscoverDescriptorsProcedure;
+    struct ReadAttributeProcedure;
+    struct ReadUsingCharacteristicUUIDProcedure;
+    struct ReadAttributeBlobProcedure;
+    struct ReadMultipleCharacteristicsProcedure;
+    struct WriteAttributeProcedure;
+    struct QueuePrepareWriteProcedure;
+    struct ExecuteWriteQueueProcedure;
+
+    template<typename ProcType, typename A0>
+    ble_error_t launch_procedure(connection_handle_t connection, const A0& a0);
+
+    template<typename ProcType, typename A0, typename A1>
+    ble_error_t launch_procedure(
+        connection_handle_t connection, const A0& a0, const A1& a1
+    );
+
+    template<typename ProcType, typename A0, typename A1, typename A2>
+    ble_error_t launch_procedure(
+        connection_handle_t connection,
+        const A0& a0, const A1& a1, const A2& a2
+    );
+
+    template<typename ProcType, typename A0, typename A1, typename A2, typename A3>
+    ble_error_t launch_procedure(
+        connection_handle_t connection,
+        const A0& a0, const A1& a1, const A2& a2, const A3& a3
+    );
+
+    GattProcedure* get_procedure(connection_handle_t) const;
+    bool register_procedure(GattProcedure*);
+    bool remove_procedure(GattProcedure*);
+
+    void handle_procedure_event(const ble_evt_t &evt);
+    void handle_hvx_event(const ble_evt_t &evt);
+    void handle_timeout_event(const ble_evt_t &evt);
+
+    static const size_t max_procedures_count =
+        CENTRAL_LINK_COUNT + PERIPHERAL_LINK_COUNT;
+
+    // Note: Ideally we would have used an array of variant here
```

Comment:
I don't understand what this comment means, is this a todo?

Reply:
Not really a todo, I won't implement a compliant variant in C++03; it requires an heavy TMP machinery. However if we have a variant type added in mbed platform then pointers can be replaced by a variant.